### DISCUSSION TITLE
grain shape and mesh size

### DISCUSSION THREAD

**Discussion by shuaizhou001** [neperfepx/neper#328] [Q&A] [answered]
Hello! 
I use this command:` neper -T -n 125 -morpho "columnar(x)"` to generate the columnar grain model (the following figure). In this model, there exists some small grains and I cannot mesh by large rcl. My purpose is to control 125  columnar grains and 3500-8000 Hex elements. Who has a method to control the random columnar grain (the grain size not have a big difference) and element number?
![columnar](https://user-images.githubusercontent.com/83024041/141866485-9c11fd46-34c3-44a3-b711-4d098698fbe5.jpg)

**Comment by smallrain1996** [comment, marked as answer]
You could try a strategy like this (with parameters that might still need to be adjusted):
```
neper -T -n 125 -morpho "size:normal(mean=0.008,sigma=0.002),columnar(x),gg"
```

For the number of elements, there is no very definite strategy and you need to keep trying

**Reply by shuaizhou001** [reply]
Thank you for your reply, This is a good method.

**Reply by rquey** [reply]
> ```
> neper -T -n 125 -morpho "size:normal(mean=0.008,sigma=0.002),columnar(x),gg"
> ```

This can give some improvements, but there are 2 problems with this command:
- `size` and `gg` are redundant (`gg` contains a `diameq` distribution, which may somewhat conflict with the `size` distribution)
- `gg` contains a `sphericity` distribution, which does not behave well for a columnar structure (the sphericity depends on the length along the columnar direction, which is not something we want...)

**Comment by rquey** [comment]
Use `-morpho` to control the grain size distribution, e.g.
```
neper -T -n 125 -morpho "diameq:lognormal(1,0.1),columnar(x)"
```
for a narrow lognormal distribution.

![image](https://user-images.githubusercontent.com/2659376/141947396-ef7d25e4-4daa-47db-90ec-4f17c2bbd50e.png)

I would generally recommend to add a sphericity distribution, but this does not behave well for elongated grains (elongated grains have a low sphericity, which depends on the length along the columnar direction...). To control the sphericity, the best is to generate a 2D tessellation and then extrude it:

```
neper -T -n 125 -morpho "diameq:lognormal(1,0.1),1-sphericity:lognormal(0.1,0.03)" -dim 2 -o 2d
neper -T -n 125 -morphooptiini "file(2d.tess)" -transform "rotate(0,1,0,90)" -o 3d
```
![image](https://user-images.githubusercontent.com/2659376/141949036-e3797eda-7698-4200-ad79-40f82757a4c2.png)
![image](https://user-images.githubusercontent.com/2659376/141948857-11f34f98-1e02-481b-896c-23b9f761ba1c.png)

**Reply by shuaizhou001** [reply]
Thank you! I have used diameq:lognormal( ) to control the size distribution and the model cannot mesh less elements. May be your second comments is good. What do you think this? 
neper -T -n 125 -morpho "size:normal(mean=0.01,sigma=0.002),columnar(x)"
![gg](https://user-images.githubusercontent.com/83024041/141951947-b5aa0a8d-6889-4f40-ab46-4d2f3f626139.png)

**Reply by rquey** [reply]
`-morpho "size:normal(mean=0.01,sigma=0.002),columnar(x)"` is fine if you don't want to enforce a sphericity distribution.

What is your meshing command?

**Reply by shuaizhou001** [reply]
neper -M n125-id1.tess -elttype hex -rcl 1.2 -format 'inp'

**Reply by rquey** [reply]
The number of elements produced by `-elttype hex` does not depend on the grain shapes or size distribution as the elements do not follow the cell boundaries. Change `rcl` to get more or less elements:

`-rcl 0.5`:
![image](https://user-images.githubusercontent.com/2659376/141954105-1fcbf5fc-e7fa-4978-a3ae-a9ecbe29fe69.png)

`-rcl 1.2`:
![image](https://user-images.githubusercontent.com/2659376/141954164-8392a587-c00f-4615-afc6-1c4539210672.png)

You may also use longer elements along x to reduce the number of elements:

`-rcl 1 -clratio 4:1:1`:
![image](https://user-images.githubusercontent.com/2659376/141954278-16372130-e18e-4205-89d4-6acd50857ea7.png)

**Reply by shuaizhou001** [reply]
Because if the model exists a small grain and use rcl 1.2, the small grain cannot mesh. The inp file will lost this grain. For this model, **rcl1.3** will mesh with a failure.

**Reply by rquey** [reply]
Il the element size is larger than a grain then, yes, some grains may not be meshed. This will happen less frequently if they are fewer small grains (small standard deviation of `size` or `diameq`).

**Reply by 123fgh123** [reply]
Dear rquey:

When I tried the following commands to generate columnar grain with periodicity, I found that 2D gains maintained periodicity very well，but the 3D gains (generated from 2D gains) lost the periodicity. 

`neper -T -n 125 -morpho "diameq:lognormal(1,0.1),1-sphericity:lognormal(0.1,0.03)" -dim 2 -periodicity all -o 2d`
![img2d](https://github.com/neperfepx/neper/assets/64315155/6b46ee79-fe61-4d5f-a3d7-9effc761d95d)

`neper -T -n 125 -morphooptiini "file(2d.tess)" -transform "rotate(0,1,0,90)" -o 3d`
![img3d](https://github.com/neperfepx/neper/assets/64315155/60ecbb18-305e-4c2a-9994-8672848bff1d)

Is there a method to generate columnar grain and maintained periodicity meanwhile?

Thank you!

Bests,
Guofeng

**Reply by rquey** [reply]
```
neper -T -n 125 -morphooptiini "file(2d.tess)" -transform "rotate(0,1,0,90)" -o 3d -periodicity x,y
```
![d](https://github.com/neperfepx/neper/assets/2659376/5fe27224-93f8-4f9c-8a71-0e83efed4163)

**Reply by 123fgh123** [reply]
Ok, thank you very much.

Is it possibile to further control the tesrsize of vtk file which was generated from columnar grain? 

I want the tesrsize to be 1:128:128, but the obtained result seemed to not be reasonable because the grain mophology was destroyed. 

`neper -T -n 125 -morphooptiini "file(2d.tess)" -transform "rotate(0,1,0,90)" -o 3d -periodicity x,y -format vtk -tesrsize 1:128:128`

![tesrsize](https://github.com/neperfepx/neper/assets/64315155/34ee4483-f268-4f31-818a-863e79065298)

Bests,
Guofeng
